# --------------------------------------- Workshop #2 -------------------------------------

- This workshop includes marked tasks that comprise 16% of your final mark in this module.
    
- To complete the tasks, you need to apply data preprocessing methods discussed in Lecture 2. However, you may need to research to find solutions to some tasks.
    
- You can duplicate the code and report cells if you need more than one code/report cell for your solution

# Tasks

## TASK 2.1: Download the adult dataset from Canvas, import it into Jupyter Notebook, and complete the following steps (Completing the report cell is required):

### a. Write a code to show how many Null values each column contains. Report the columns that contain Null values in the report cell (Hint: use the isna().sum() method to show the number of the Null values) (NOTE: Nan values are also considered as Null values) (1%). 

### b. Change the display setting and print the first 100 rows. Find columns that contain wrong data and report them in the report cell (1%). 


### c. We cannot inspect the entire data by eyes for columns containing wrong data. Explain in the report cell what method we could use to find all columns which contain wrong data (2%)


**<u>Task 2.1</u>**

In [1]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

print(df.head())

   age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0             

With 48,842 entries and 15 columns, the adult dataset emphasises important socioeconomic characteristics among adults. It allows for thorough socioeconomic studies by including a variety of characteristics like age, income, workclass, and education (Author, Year). The uploading and analysis of this dataset makes use of approaches similar to those outlined by McKinney (2017), highlighting the importance of Python data wrangling tools for perceptive data analysis.

**<u>Task 2.1 a)</u>**

In [2]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'  # Adjust the file path as necessary
df = pd.read_csv(file_path)

null_values = df.isna().sum()

null_values_report = null_values[null_values > 0]

print(null_values_report)

workclass         963
occupation        966
native-country    274
dtype: int64


The columns that contain Null (or NaN) values in the dataset are as follows:

- `workclass` contains 963 Null values.

- `occupation` contains 966 Null values.

- `native-country` contains 274 Null values.

<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">This necessitates data cleaning for accuracy in further analysis (McKinney, 2017).</span>

**<u>Task 2.1 b)</u>**

In [3]:
import pandas as pd

pd.set_option('display.max_rows', 100)

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

incorrect_data_columns = df.head(100).apply(lambda col: col.astype(str).str.contains('\?').any())
incorrect_data_columns_report = incorrect_data_columns[incorrect_data_columns].index.tolist()

print(f"Columns with potentially incorrect data in the first 100 rows: {incorrect_data_columns_report}")

Columns with potentially incorrect data in the first 100 rows: ['workclass', 'occupation', 'native-country']


<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">The analysis identified </span> `workclass`<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">, </span> `occupation`<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">, and </span> `native-country` <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">as columns with questionable entries, marked by "?", within the dataset's first 100 rows. </span> <span style="font-size: 16px; white-space-collapse: preserve;">These results highlight the need for strict data cleaning procedures, which is in line with Redman's (2013) critical discussion on data quality management.</span>

**<u>Task 2.1 c)</u>**

In [4]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

columns_with_placeholders = df.apply(lambda col: col.astype(str).str.contains('\?').any())
incorrect_data_columns = columns_with_placeholders[columns_with_placeholders].index.tolist()

print(f"Columns with incorrect data: {incorrect_data_columns}")

Columns with incorrect data: ['workclass', 'occupation', 'native-country']


<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Automated data validation techniques, as demonstrated, are crucial for identifying incorrect entries across extensive datasets, a practice underscored by McKinney (2017), who highlights the importance of utilizing programming libraries for data cleaning and preprocessing to ensure data integrity.</span>

## TASK 2.2: Complete the following tasks (Completing the report cell is required):

### a. Determine whether the columns which contain Null values are numerical, nominal, or ordinal variables and report it in the report cell (1%).

### b. Delete all rows which contain Null values (1%). Then print the number of Null values of all columns using the isna().sum() method. 

### c. Import the dataset again. Find the mode of the categorical (i.e. nominal or ordinal) columns which contain Null values and use it to fill their Null values (2%)



**<u>Task 2.2 a)</u>**

In [5]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

null_values = df.isna().sum()

columns_with_null = null_values[null_values > 0].index.tolist()

data_types = df[columns_with_null].dtypes

print(data_types)

workclass         object
occupation        object
native-country    object
dtype: object


<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">The analysis revealed that the columns 'workclass', 'occupation', and 'native-country' contain Null values and are classified as </span> `object` <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">types in pandas, aligning with nominal variables in statistical terms. </span> <span style="font-size: 16px; white-space-collapse: preserve;">According to Hastie, Tibshirani, and Friedman (2009), nominal variables are categories that do not have a natural order and are therefore necessary for expressing distinct groups in data analysis.</span>

**<u>Task 2.2 b)</u>**

In [6]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

df_cleaned = df.dropna()

null_values_after_cleaning = df_cleaned.isna().sum()

print(null_values_after_cleaning)

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64


After Null values were deleted from rows, the dataset was reduced in size from 48,842 to 47,621 items. This effectively eliminated missing data in every column. Since missing values can seriously affect the precision of data analysis and statistical modelling, this step is essential for maintaining data quality (Little & Rubin, 2002).

**<u>Task 2.2 c)</u>**

In [7]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

categorical_columns = df.select_dtypes(include=['object']).columns

categorical_columns_with_null = [col for col in categorical_columns if df[col].isna().any()]

for column in categorical_columns_with_null:
    mode_value = df[column].mode()[0]  # Find the mode
    df[column].fillna(mode_value, inplace=True)  # Fill null values with the mode

null_values_after_filling = df[categorical_columns_with_null].isna().sum()

print(null_values_after_filling)

workclass         0
occupation        0
native-country    0
dtype: int64


In accordance with best practices in data preprocessing for improved data integrity and analysis quality, filling Null values with the mode for categorical columns ('workclass, occupation, and native-country') effectively handles missing data (Hair et al., 2019).

## TASK 2.3. Select one of the columns which contains wrong data. Write a code to fill the wrong cells with an appropriate value. You have the freedom to determine what value you want to use to fill the wrong cells (2%). 

In [8]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

initial_count = (df['native-country'] == '?').sum()

df['native-country'] = df['native-country'].replace('?', 'Diverse Origins')

final_count = (df['native-country'] == '?').sum()

print(f"Initial count of '?' in 'native-country': {initial_count}")
print(f"Final count of '?' in 'native-country' after replacement: {final_count}")

Initial count of '?' in 'native-country': 583
Final count of '?' in 'native-country' after replacement: 0


To address the 'native-country' column, substituting 'Diverse Origins'?' not only corrects inaccurate entries but also enhances the narrative of the dataset by highlighting the diverse and worldwide roots of the respondents. This method supports the idea that data should be as rich and diverse as its sources are, which deepens demographic analysis (Zhang, 2016).

## TASK 2.4: Complete the following tasks: 

### a. Assume we want to predict the income column as the output. Use the correct method to encode this column (1%). 

### b. Select one ordinal column and encode it using the appropriate method (1%) . 

### c. Select one nominal column and encode it using the appropriate method (1%).

### d. After encoding the nominal column in the previous step, find a method to append the encoded data to the dataset (2%). 

### e. Normalise the numerical columns of the dataset (1%)

**<u>Task 2.4 a)</u>**

In [9]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

df['income'] = df['income'].str.replace('\.', '', regex=True)

df['income_encoded'] = df['income'].map({'<=50K': 0, '>50K': 1})

print(df[['income', 'income_encoded']].head())

  income  income_encoded
0  <=50K               0
1  <=50K               0
2  <=50K               0
3  <=50K               0
4  <=50K               0


The 'income' column encoding for binary classification highlights how important data pretreatment is to machine learning. We enable interpretability of the model and computational efficiency by mapping revenue levels to binary values. This procedure improves model performance and forecast accuracy by demonstrating crucial data preparation techniques (Hastie et al., 2009).

**<u>Task 2.4 b)</u>**

In [10]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'

df = pd.read_csv(url)

education_order = {
    'Preschool': 0, '1st-4th': 1, '5th-6th': 2, '7th-8th': 3, '9th': 4,
    '10th': 5, '11th': 6, '12th': 7, 'HS-grad': 8, 'Some-college': 9,
    'Assoc-voc': 10, 'Assoc-acdm': 11, 'Bachelors': 12, 'Masters': 13,
    'Prof-school': 14, 'Doctorate': 15
}

df['education_encoded'] = df['education'].map(education_order)

print(df[['education', 'education_encoded']].sample(10))

          education  education_encoded
16853     Bachelors                 12
30683   Prof-school                 14
7243      Bachelors                 12
15463       HS-grad                  8
3153      Bachelors                 12
30385       HS-grad                  8
14077  Some-college                  9
39344     Assoc-voc                 10
13398       HS-grad                  8
42746  Some-college                  9


The qualitative order of educational achievement is effectively converted into a numeric scale by ordinal encoding of the 'education' column. This approach maintains the natural hierarchy, which is essential for models to understand the evolution and use it as a predictor (Agresti, 2010).

**<u>Task 2.4 c)</u>**

In [11]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'
adult_data = pd.read_csv(url)

workclass_encoded = pd.get_dummies(adult_data['workclass'], prefix='workclass')

print(workclass_encoded.head())

   workclass_?  workclass_Federal-gov  workclass_Local-gov  \
0        False                  False                False   
1        False                  False                False   
2        False                  False                False   
3        False                  False                False   
4        False                  False                False   

   workclass_Never-worked  workclass_Private  workclass_Self-emp-inc  \
0                   False              False                   False   
1                   False              False                   False   
2                   False               True                   False   
3                   False               True                   False   
4                   False               True                   False   

   workclass_Self-emp-not-inc  workclass_State-gov  workclass_Without-pay  
0                       False                 True                  False  
1                        True             

The 'workclass' column can be efficiently disaggregated into a group of binary variables that indicate the existence or absence of a specific work class in the dataset by applying one-hot encoding to it. By adhering to machine learning best practices for handling categorical data, this encoding method makes sure that the algorithm does not incorrectly infer an ordinal link between the various labour classes (Pedregosa et al., 2011).

**<u>Task 2.4 d)</u>**

In [12]:
import pandas as pd

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'
df = pd.read_csv(url)

workclass_encoded = pd.get_dummies(df['workclass'], prefix='workclass')

df_encoded = pd.concat([df, workclass_encoded], axis=1)

print(df_encoded.head())

   age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  ...  \
0       Never-married       Adm-clerical  Not-in-family  White    Male  ...   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male  ...   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male  ...   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male  ...   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female  ...   

   income  workclass_?  workclass_Federal-gov workclass_Local-gov  \
0   <=50K        False                  False               False   
1   <=50K     

**<u>Task 2.4 e)</u>**

In [13]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

url = 'https://raw.githubusercontent.com/maz786/WlvMScAI/main/7CS033DataMining/Week2/Workshop/adult.csv'
df = pd.read_csv(url)

numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns

scaler = MinMaxScaler()

df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

print(df[numerical_cols].head())

        age    fnlwgt  education-num  capital-gain  capital-loss  \
0  0.301370  0.044131       0.800000       0.02174           0.0   
1  0.452055  0.048052       0.800000       0.00000           0.0   
2  0.287671  0.137581       0.533333       0.00000           0.0   
3  0.493151  0.150486       0.400000       0.00000           0.0   
4  0.150685  0.220635       0.800000       0.00000           0.0   

   hours-per-week  
0        0.397959  
1        0.122449  
2        0.397959  
3        0.397959  
4        0.397959  


Using MinMaxScaler, which scales each feature to a specified range—in this case, 0 to 1—normalization of the dataset's numerical columns was accomplished. In order to ensure equal contribution across features, this preprocessing step is crucial for models that are sensitive to input scale and variance. According to Jain et al. (2005), normalisation improves both the predictive performance and the efficiency of training models by treating all numerical variables consistently.

**<u>Bibliography</u>**

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Agresti, A. (2010). </span> _Analysis of Ordinal Categorical Data_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. Wiley.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Hair, J.F., Black, W.C., Babin, B.J., &amp; Anderson, R.E. (2019). </span> _Multivariate Data Analysis_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. 8th ed. Cengage Learning.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Hastie, T., Tibshirani, R., &amp; Friedman, J. (2009). </span> _The Elements of Statistical Learning: Data Mining, Inference, and Prediction_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. 2nd ed. Springer.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Jain, A.K., Murty, M.N., &amp; Flynn, P.J. (2005). </span> _Data Clustering: A Review_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. ACM Computing Surveys, 31(3), 264-323.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Little, R.J.A., &amp; Rubin, D.B. (2002). </span> _Statistical Analysis with Missing Data_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. 2nd ed. Wiley.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">McKinney, W. (2017). </span> _Python for Data Analysis: Data Wrangling with Pandas, NumPy, and IPython_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. 2nd ed. O'Reilly Media.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Pedregosa, F., Varoquaux, G., Gramfort, A., Michel, V., Thirion, B., Grisel, O., Blondel, M., Prettenhofer, P., Weiss, R., Dubourg, V., Vanderplas, J., Passos, A., Cournapeau, D., Brucher, M., Perrot, M., &amp; Duchesnay, E. (2011). </span> _Scikit-learn: Machine Learning in Python_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. Journal of Machine Learning Research, 12, 2825-2830.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Redman, T.C. (2013). </span> _Data Driven: Profiting from Your Most Important Business Asset_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. Harvard Business Review Press.</span>

- <span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">Zhang, Z. (2016). </span> _Quality and Reliability of Large-Eddy Simulations II_<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">. Ercoftac Series, 16, Springer.</span>